autoload/coverage/gcov/parsing.vim

```diff
@@ -0,0 +1,135 @@
+" Copyright 2021 Google Inc. All rights reserved.
+"
+" Licensed under the Apache License, Version 2.0 (the "License");
+" you may not use this file except in compliance with the License.
+" You may obtain a copy of the License at
+"
+"     http://www.apache.org/licenses/LICENSE-2.0
+"
+" Unless required by applicable law or agreed to in writing, software
+" distributed under the License is distributed on an "AS IS" BASIS,
+" WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+" See the License for the specific language governing permissions and
+" limitations under the License.
+
+"{{{ Init
+
+let s:plugin = maktaba#plugin#Get('coverage')
+
+"}}}
+
+"{{{ Utility functions for parsing gcov files (i.e. lcov tracefiles).
+
+""
+" @private
+" Categorizes a line potentially containing coverage information in a coverage
+" file.  Recognized coverage information is be formatted as:
+" - DA:<line number>,<execution count>[,<checksum>]
+" - BA:<line number>,<branch coverage (0: uncovered, 1: partial, 2: covered>
+" - BRDA:<line number>,<block number>,<branch number>,<taken>
+"
+" Note that for BRDA lines, we can't check for partial coverage (only '-' for
+" uncovered, or 1+ for the number of times the line was run).
+"
+" For additional details on lcov's tracefile format, see FILES under on
+" geninfo's man page, or view the man page at:
+" http://ltp.sourceforge.net/coverage/lcov/geninfo.1.php
+function! s:TryParseLine(line) abort
+  let [l:prefix, l:raw_info] = split(a:line, ':', 1)
+  let l:prefix = maktaba#string#Strip(l:prefix)
+
+  if l:prefix ==? 'BA' || l:prefix ==? 'DA'
+    let l:hits_index = 1
+  elseif l:prefix ==? 'BRDA'
+    let l:hits_index = 3
+  else
+    return []
+  endif
+
+  try
+    let l:info = split(l:raw_info, ',', 1)
+    let l:linenum = str2nr(info[0])
+    " Note that For BRDA lines, '-' is used instead of '0' for uncovered, which
+    " str2nr will safely convert to 0.
+    let l:hits = str2nr(info[l:hits_index])
+  catch
+    call s:plugin.logger.Debug(
+          \ 'Failed to parse gcov line (%s): %s', v:exception, a:line)
+    return []
+  endtry
+
+  if hits is 0
+    return ['uncovered', l:linenum]
+  endif
+
+  if hits is 1 && l:prefix ==? 'BA'
+    return ['partial', l:linenum]
+  endif
+
+  return ['covered', l:linenum]
+endfunction
+
+""
+" @private
+" Gets a list of covered filenames and reports for a given lcov tracefile.
```

Comment:
Could use an explicit example of the format of data it returns.

Is it guaranteed to have "covered", "partial" and "uncovered" for each filename? From the code I can't see how it would be guaranteed to include each category, but helpers like `s:ExtendReport` seem to assume it will be and would blow up if not.

Reply:
I think it should be safe, because the function:
- calls coverage#CreateReport when it enters a file record, which returns a dict containing all three.
- handles being told to write a record with no SF (start of file) line
- handles being told to write a record for an empty report

I've added some docs, and tests to verify, but lmk if you see something I'm missing.

Reply:
I do see a place this could be better.  In `coverage#GetFormattedStats`, we assume that total > 0, and can end up printing "Coverage data is nan%".

In the case of gcov, it is possible to have an empty report for a file, such as:

```
SF:foo.c
end_of_report
```

this will produce:

```
{'covered': [], 'uncovered': [], partial: []}
```

The only place I can see this being an issue is in `coverage#GetFormattedStats`, so I've added a guard clause there to display a better message in that case, but lmk if you'd like something different.